File: idakonyvtar/src/main/java/eu/eyan/idakonyvtar/util/DialogHelper.java
package eu.eyan.idakonyvtar.util;

import java.awt.BorderLayout;
import java.awt.Component;
import java.awt.Dimension;
import java.awt.Toolkit;
import java.awt.Window;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;

import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JMenuBar;
import javax.swing.JOptionPane;
import javax.swing.JScrollPane;
import javax.swing.JToolBar;
import javax.swing.SwingUtilities;

import org.jdesktop.swingx.JXFrame;

import com.jgoodies.forms.builder.PanelBuilder;
import com.jgoodies.forms.factories.CC;
import com.jgoodies.forms.layout.FormLayout;

import eu.eyan.idakonyvtar.controller.IController;
import eu.eyan.idakonyvtar.controller.IControllerWithMenu;
import eu.eyan.idakonyvtar.controller.IDialogController;

public class DialogHelper
{

    public static final String CANCEL = "Mégsem";
    public static final String SAVE = "Mentés";

    /**
     * Blockiert!!!
     * 
     * @param parent
     * @param controller
     * @param input
     * @return
     */
    public static <INPUT, OUTPUT> OkCancelDialog startModalDialog(Component parent, IDialogController<INPUT, OUTPUT> controller, INPUT input)
    {
        controller.initData(input);

        Window parentWindow = parent == null ? null : SwingUtilities.windowForComponent(parent);
        OkCancelDialog dialog = new OkCancelDialog(parentWindow);
        dialog.setLocationRelativeTo(parent);
        dialog.setModal(true);
        PanelBuilder panelBuilder = new PanelBuilder(new FormLayout("pref:grow", "top:pref:grow, 3dlu, pref"));
        panelBuilder.add(controller.getView(), CC.xy(1, 1));
        panelBuilder.add(getButtons(dialog, controller), CC.xy(1, 3));
        dialog.add(addScrollableInBorders(panelBuilder.build()));
        dialog.setTitle(controller.getTitle());
        dialog.setResizable(true);
        dialog.addWindowListener(new WindowAdapter()
        {
            @Override
            public void windowClosing(WindowEvent e)
            {
                super.windowClosed(e);
            }
        });

        controller.initBindings();
        dialog.pack();
        positionToCenter(dialog);
        controller.addResizeListener(dialog);

        // blockiert:
        dialog.setVisible(true);
        if (parentWindow != null)
        {
            parentWindow.invalidate();
        }
        return dialog;
    }

    private static void positionToCenter(Component component)
    {
        Dimension screenSize = Toolkit.getDefaultToolkit().getScreenSize();
        int width = component.getSize().width;
        int height = component.getSize().height;
        component.setSize(width, height);
        component.setLocation((screenSize.width - width) / 2, (screenSize.height - height) / 2);
    }

    private static Component addScrollableInBorders(Component component)
    {
        PanelBuilder panelBuilder = new PanelBuilder(new FormLayout("3dlu,pref:grow,3dlu", "3dlu,pref:grow,3dlu"));
        panelBuilder.add(component, CC.xy(2, 2));
        JScrollPane scrollPane = new JScrollPane(panelBuilder.build());
        scrollPane.setBorder(null);
        return scrollPane;
    }

    private static <INPUT> JFrame runInFrame(Component parent, IController<INPUT, ?> controller, INPUT input, JMenuBar jMenuBar, JToolBar toolBar, boolean fullScreen)
    {
        controller.initData(input);
        JXFrame frame = new JXFrame();
        frame.add(toolBar, BorderLayout.NORTH);
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.add(controller.getView());
        frame.setTitle(controller.getTitle());
        frame.setName(controller.getTitle());
        frame.setJMenuBar(jMenuBar);
        controller.initBindings();
        frame.setVisible(true);
        frame.pack();
        positionToCenter(frame);
        if (fullScreen)
        {
            frame.setExtendedState(JFrame.MAXIMIZED_BOTH);
        }
        Component initFocusComponent = controller.getComponentForFocus();
        if (initFocusComponent != null)
        {
            initFocusComponent.requestFocusInWindow();
        }
        return frame;
    }

    private static Component getButtons(final OkCancelDialog dialog, final IDialogController<?, ?> dialogController)
    {
        PanelBuilder panelBuilder = new PanelBuilder(new FormLayout("pref:grow,3dlu,pref,3dlu,pref", "pref"));
        JButton okButton = new JButton(SAVE);
        okButton.setName(SAVE);
        JButton cancelButton = new JButton(CANCEL);
        cancelButton.setName(CANCEL);
        panelBuilder.add(okButton, CC.xy(3, 1));
        panelBuilder.add(cancelButton, CC.xy(5, 1));
        okButton.addActionListener(new ActionListener()
        {
            @Override
            public void actionPerformed(ActionEvent e)
            {
                dialogController.onOk();
                dialog.ok = true;
                dialog.dispose();
            }
        });
        cancelButton.addActionListener(new ActionListener()
        {
            @Override
            public void actionPerformed(ActionEvent e)
            {
                dialogController.onCancel();
                dialog.dispose();
            }
        });
        return panelBuilder.build();
    }

    public static <INPUT> JFrame runInFrameFullScreen(IControllerWithMenu<INPUT, ?> controller, INPUT input)
    {
        return runInFrame(null, controller, input, controller.getMenuBar(), controller.getToolBar(), true);
    }

    public static boolean yesNo(Component parent, String question, String dialogTitle)
    {
        return JOptionPane.showOptionDialog(parent,
                question,
                dialogTitle,
                JOptionPane.YES_NO_OPTION,
                JOptionPane.QUESTION_MESSAGE,
                null,
                new Object[] { "Igen", "Mégsem" },
                "Mégsem") == JOptionPane.OK_OPTION;
    }
}

File: idakonyvtar/src/main/scala/eu/eyan/idakonyvtar/util/DialogHelper.scala
package eu.eyan.idakonyvtar.util;

import java.awt.BorderLayout
import java.awt.Component
import java.awt.Frame
import java.awt.Toolkit
import java.awt.Window
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import org.jdesktop.swingx.JXFrame
import com.jgoodies.forms.builder.PanelBuilder
import com.jgoodies.forms.factories.CC
import com.jgoodies.forms.layout.FormLayout
import eu.eyan.idakonyvtar.controller.IController
import eu.eyan.idakonyvtar.controller.IControllerWithMenu
import eu.eyan.idakonyvtar.controller.IDialogController
import javax.swing.JButton
import javax.swing.JFrame
import javax.swing.JMenuBar
import javax.swing.JOptionPane
import javax.swing.JScrollPane
import javax.swing.JToolBar
import javax.swing.SwingUtilities
import eu.eyan.util.swing.OkCancelDialog
import eu.eyan.util.swing.OkCancelDialog
import eu.eyan.util.swing.JPanelWithFrameLayout
import javax.swing.JPanel
import eu.eyan.util.awt.ComponentPlus.ComponentPlusImplicit
import eu.eyan.util.swing.JDialogPlus.JdialogPlusImplicit
import eu.eyan.util.text.Text
import eu.eyan.util.swing.JButtonPlus.JButtonImplicit
import eu.eyan.util.swing.Alert
import javax.swing.UIManager
import javax.swing.JFileChooser
import eu.eyan.util.swing.JFileChooserPlus.JFileChooserImplicit
import java.io.File
import eu.eyan.util.text.TextsDialogYes
import eu.eyan.util.text.TextsDialogYesNo
import eu.eyan.util.text.TextsDialogFileChooser
import eu.eyan.util.text.TextsDialogYesNoCancel
import eu.eyan.util.swing.JDialogPlus.JdialogPlusImplicit
import java.awt.GraphicsEnvironment
import eu.eyan.util.swing.JPanelPlus.JPanelImplicit
import eu.eyan.util.awt.ContainerPlus.ContainerPlusImplicit
import eu.eyan.util.swing.JComponentPlus.JComponentImplicit
import eu.eyan.idakonyvtar.text.TechnicalTextsIda._
import eu.eyan.util.swing.WithComponent
import javax.swing.JDialog

object DialogHelper {

  private class JDialogWithYesNoButtons(owner: Component, content: Component, yesText: Text, noText: Text) extends JDialog(owner.windowForComponent) {
    def hasYesSelected = clickOnYes

    private var clickOnYes = false
    private val yesNoButtonsPanel = new JPanelWithFrameLayout().withSeparators
    yesNoButtonsPanel.newColumn.addButton(yesText).name(SAVE_BUTTON_NAME).onAction { clickOnYes = true; this.dispose }
    yesNoButtonsPanel.newColumn.addButton(noText).name(CANCEL_BUTTON_NAME).onAction(this.dispose)

    private val buttonsAndContent = new JPanelWithFrameLayout().withSeparators.withBorders
      .newColumnScrollable
      .newRow.addFluent(yesNoButtonsPanel)
      .newRowScrollable.addFluentInScrollPane(content)

    add(buttonsAndContent)
  }

  /** Blocks as a modal dialog!!! */
  def yesNoEditor(owner: Component, content: Component, title: Text, yesText: Text, noText: Text) =
    new JDialogWithYesNoButtons(owner, content, yesText, noText)
      .locationRelativeTo(owner)
      .modal
      .title(title)
      .resizeable
      .packFluent
      .positionToCenter
      .maximize
      .visible // blocks !!! for jdialog
      .hasYesSelected

  def yes(texts: TextsDialogYes): Option[String] = yes(texts.text, texts.title, texts.yes)
  def yes(text: Text, title: Text, yes: Text) = Alert.alertOptions(title.get, text.get, Array(yes.get))

  def yesNo(parent: Component, texts: TextsDialogYesNo): Boolean = yesNo(parent, texts.text, texts.title, texts.yes, texts.no)
  def yesNo(parent: Component, question: Text, dialogTitle: Text, yes: Text, no: Text) =
    Text.combineMoreTextsWithNamesAndExecute("question" -> question, "title" -> dialogTitle, "yes" -> yes, "no" -> no)(texts =>
      JOptionPane.showOptionDialog(
        parent,
        texts("question"),
        texts("title"),
        JOptionPane.YES_NO_OPTION,
        JOptionPane.QUESTION_MESSAGE,
        null,
        Array[Object](texts("yes"), texts("no")),
        texts("no")) == JOptionPane.OK_OPTION)

  trait DialogResult
  case object YES extends DialogResult
  case object NO extends DialogResult
  case object CANCEL extends DialogResult

  def yesNoCancel(parent: Component, texts: TextsDialogYesNoCancel): DialogResult = yesNoCancel(parent, texts.text, texts.title, texts.yes, texts.no, texts.cancel)
  def yesNoCancel(parent: Component, text: Text, title: Text, yes: Text, no: Text, cancel: Text): DialogResult =
    Text.combineMoreTextsWithNamesAndExecute("question" -> text, "title" -> title, "yes" -> yes, "no" -> no, "cancel" -> cancel)(texts => {
      val result = JOptionPane.showOptionDialog(
        parent,
        texts("question"),
        texts("title"),
        JOptionPane.YES_NO_CANCEL_OPTION,
        JOptionPane.QUESTION_MESSAGE,
        null,
        Array[Object](texts("yes"), texts("no"), texts("cancel")),
        texts("cancel"))
      if (result == JOptionPane.OK_OPTION) YES
      else if (result == JOptionPane.NO_OPTION) NO
      else CANCEL
    })

  def fileChooser(parent: Component, currentDir: File, fileFilter: String, texts: TextsDialogFileChooser, action: File => Unit) = {
    UIManager.put("FileChooser.cancelButtonText", texts.cancel.get)
    new JFileChooser()
      .withCurrentDirectory(currentDir)
      .withDialogTitle(texts.title.get)
      .withApproveButtonText(texts.approve.get)
      .withFileFilter(fileFilter, texts.fileFilterText.get)
      .showAndHandleResult(parent, action)
  }
}